src/main/java/org/reactome/addlinks/dataretrieval/ensembl/EnsemblBioMartRetriever.java

```diff
@@ -0,0 +1,247 @@
+            }
+        }
+
+        // Write data to file.
+        for (String biomartIdentifierLine : biomartIdentifierLines) {
+            Files.write(
+                    Paths.get(biomartFilename),
+                    biomartIdentifierLine.getBytes(),
+                    StandardOpenOption.APPEND, StandardOpenOption.CREATE
+            );
+
+        }
+    }
+
+    // Checks that the 4th column contains identifiers (either Microarray or UniProt ).
+    private boolean lineContainsIdentifier(List<String> tabSplit) {
+        final int IDENTIFIER_COLUMN = 3;
+        return tabSplit.size() > IDENTIFIER_COLUMN && !tabSplit.get(IDENTIFIER_COLUMN).trim().isEmpty();
+    }
+
+
+    /**
+     * This method queries BioMart using either a URL (see variable 'microarrayTypesBaseQuery) or an XML (see biomart-query.xml in resources) query.
+     * It will retry up to 5 times if errors are returned from BioMart instead of data.
+     * There are multiple cases (Frog, Both Yeasts, P. falciparum and D. discoideum) where the data does not exist in BioMart.
+     * If/When this exception is thrown for this species, it can be ignored. Check the logs from previous runs to confirm.
+     * @param queryString - String, URL/XML string that will be used to query BioMart.
+     * @param retryCount - int, Denotes how many times this query has been tried with BioMart.
+     * @return - Set<String>, All lines of successful BioMart response.
+     * @throws IOException - Thrown by HttpURLConnection, BufferedReader, URL classes.
+     * @throws InterruptedException - Thrown if Sleep is interrupted when waiting to retry BioMart query.
+     * @throws BioMartQueryException - Thrown if BioMart query doesn't match any existing data in their database.
+     * @throws HttpException - Thrown when the Http request to BioMart returns a non-200 response.
+     */
+    private Set<String> queryBioMart(String queryString, int retryCount) throws IOException, InterruptedException, BioMartQueryException, HttpException {
+        final int MAX_QUERY_RETRIES = 5;
+        // Create connection to BioMart URL for each species, retrieving a list of microarray probe types, if available.
+        URL biomartUrlWithSpecies = new URL(queryString);
+        HttpURLConnection biomartConnection = (HttpURLConnection) biomartUrlWithSpecies.openConnection();
+        if (biomartConnection.getResponseCode() == HttpStatus.SC_OK) {
+            Set<String> biomartResponseLines = new HashSet<>();
+            BufferedReader br = new BufferedReader(new InputStreamReader(biomartConnection.getInputStream()));
+            String line;
+            while ((line = br.readLine()) != null) {
+                // BioMart still responds with a 200, even if no data exists. For now, we handle it by
+                // checking the returned content for the string 'ERROR'. It will retry up to 5 times, with a 10 second delay.
+                if (line.contains("ERROR") ) {
+                    retryCount++;
```

Comment:
Modifying the value of a parameter from inside a function is usually something best avoided if possible.

Reply:
Done